bindings/java/java/src/main/java/org/ethereum/evmc/Host.java

```diff
@@ -77,12 +77,12 @@ static ByteBuffer copy_code(int context_index, byte[] address, int code_offset)
             "HostContext does not exist for context_index: " + context_index);
     byte[] code = context.getCode(address).array();
 
-    if (code != null && code_offset > 0 && code_offset < code.length) {
+    if (code_offset < code.length) {
       int length = code.length - code_offset;
-      return ByteBuffer.wrap(code, 0, length);
+      return ByteBuffer.allocateDirect(length).put(code, code_offset, length);
```

Comment:
Remove wrap - that uses a heap buffer. Use a direct buffer instead.

Reply:
It seems this is more like a bugfix, because `wrap:
```
The new buffer will be backed by the given byte array; that is, modifications to the buffer will cause the array to be modified and vice versa. The new buffer's capacity will be array.length, its position will be offset, its limit will be offset + length, and its mark will be undefined. Its backing array will be the given array, and its array offset will be zero.
```

And unless `getCode` returns a new copy, this would mean the underlying code would be modifiable. Thought the `wrap` code is definitely more efficient in the sense it does no copying. Really need to verify what `getCode` does.

Can you pull out this change into a PR of its own?

Ref:
- https://docs.oracle.com/javase/7/docs/api/java/nio/ByteBuffer.html#wrap(byte[],%20int,%20int)
- https://docs.oracle.com/javase/7/docs/api/java/nio/ByteBuffer.html#allocateDirect(int)
- https://docs.oracle.com/javase/7/docs/api/java/nio/ByteBuffer.html#put(byte[],%20int,%20int)

Reply:
The docs of `.getCode()` says

> The Client MUST copy the requested code...

So changing wrap (no-copy) to copy looks good to me.

I'm also fine with leaving the fix here because this PR is "Essential fixes for EVMC Java bindings".

Reply:
right - I found out in testing that if I wrap and send to native C we get a crash. By using a direct buffer, we fix the issue.

Reply:
I'd argue either it is split out or this PR is made into having separate functional changes in separate commits. It is hard to review in its current form, especially as it reformats the C file.

Reply:
Separated here: https://github.com/ethereum/evmc/pull/535

Reply:
Meged :wink: